go/vt/wrangler/tablet_test.go

```diff
@@ -56,3 +58,102 @@ func TestInitTabletShardConversion(t *testing.T) {
 	}
 }
+
+// TestDeleteTabletBasic tests delete of non-master tablet
+func TestDeleteTabletBasic(t *testing.T) {
+	cell := "cell1"
+	ts := memorytopo.NewServer(cell)
+	wr := New(logutil.NewConsoleLogger(), ts, nil)
+
+	tablet := &topodatapb.Tablet{
+		Alias: &topodatapb.TabletAlias{
+			Cell: cell,
+			Uid:  1,
+		},
+		Shard: "0",
+	}
+
+	if err := wr.InitTablet(context.Background(), tablet, false /*allowMasterOverride*/, true /*createShardAndKeyspace*/, false /*allowUpdate*/); err != nil {
+		t.Fatalf("InitTablet failed: %v", err)
+	}
+
+	if _, err := ts.GetTablet(context.Background(), tablet.Alias); err != nil {
+		t.Fatalf("GetTablet failed: %v", err)
+	}
+
+	if err := wr.DeleteTablet(context.Background(), tablet.Alias, false); err != nil {
+		t.Fatalf("DeleteTablet failed: %v", err)
+	}
+}
+
+// TestDeleteTabletTrueMaster tests that you can delete a true master tablet
+// only if allowMaster is set to true
+func TestDeleteTabletTrueMaster(t *testing.T) {
+	cell := "cell1"
+	ts := memorytopo.NewServer(cell)
+	wr := New(logutil.NewConsoleLogger(), ts, nil)
+
+	_, err := ts.GetOrCreateShard(context.Background(), "test", "0")
+	tablet := &topodatapb.Tablet{
+		Alias: &topodatapb.TabletAlias{
+			Cell: cell,
+			Uid:  1,
+		},
+		Keyspace: "test",
+		Shard:    "0",
+		Type:     topodatapb.TabletType_MASTER,
+	}
+
+	if err := wr.InitTablet(context.Background(), tablet, false /*allowMasterOverride*/, false /*createShardAndKeyspace*/, false /*allowUpdate*/); err != nil {
+		t.Fatalf("InitTablet failed: %v", err)
+	}
+	if _, err := ts.GetTablet(context.Background(), tablet.Alias); err != nil {
+		t.Fatalf("GetTablet failed: %v", err)
+	}
+
+	// set MasterAlias and MasterTermStartTime on shard to match chosen master tablet
+	_, err = ts.UpdateShardFields(context.Background(), "test", "0", func(si *topo.ShardInfo) error {
+		si.MasterAlias = tablet.Alias
+		si.MasterTermStartTime = tablet.MasterTermStartTime
+		return nil
+	})
+
+	err = wr.DeleteTablet(context.Background(), tablet.Alias, false)
+	wantError := "as it is a master, use allow_master flag"
+	if err == nil || !strings.Contains(err.Error(), wantError) {
+		t.Fatalf("DeleteTablet on master: want error = %v, got error = %v", wantError, err)
+	}
+
+	if err := wr.DeleteTablet(context.Background(), tablet.Alias, true); err != nil {
+		t.Fatalf("DeleteTablet failed: %v", err)
+	}
+}
+
+// TestDeleteTabletFalseMaster tests that you can delete a false master tablet
+// with allowMaster set to false
+func TestDeleteTabletFalseMaster(t *testing.T) {
+	cell := "cell1"
+	ts := memorytopo.NewServer(cell)
+	wr := New(logutil.NewConsoleLogger(), ts, nil)
+
+	tablet := &topodatapb.Tablet{
+		Alias: &topodatapb.TabletAlias{
+			Cell: cell,
+			Uid:  1,
+		},
+		Keyspace: "test",
+		Shard:    "0",
+		Type:     topodatapb.TabletType_MASTER,
+	}
+
+	if err := wr.InitTablet(context.Background(), tablet, false /*allowMasterOverride*/, true /*createShardAndKeyspace*/, false /*allowUpdate*/); err != nil {
```

Comment:
I'm a bit confused how this test is working. Shouldn't init tablet fail here if there is already a master tablet? 

https://github.com/vitessio/vitess/blob/510b1fe379415d7a5a078c47dc80d8bc3233e7d1/go/vt/wrangler/tablet.go#L71

Reply:
`HasMaster()` will be false (si.MasterAlias is nil until it is set).

Reply:
I have modified the test case to more closely reflect the real-life use case.